Refactor & update fooof.synth.transform

TomDonoghue opened the pull request:
Responds to / related to #109

This is a primarily a refactor and extension of fooof.synth.transform. A couple other small updates for naming consistency (using the same convention for gauss*_params between SynParams & gen, and standardizing on calling the ap param 'offset', instead of 'intercept'). 

Notes / comments (mostly Qs for @rdgao):
- Do we know how the knee behaves with these kind of transforms? From quick tests, it appears that, rotating the spectrum does not preserve the knee (which, I think, is to be expected). I'm guessing we can work through figuring out how it does so - though perhaps for now we can just add a note that this is only designed for rotating 'fixed' spectra? (I don't currently need the knee case). 
- I switched the convention for the direction of rotation, so now positive value steepens, etc (this, I think, is consistent with FOOOFs conventions of using positive exponents). This seems like the more obvious way to do it (given how FOOOF sets up the equations), since a delta of + 0.2 can be interpreted as "OG-val + 0.2", and -0.2 as "OG-val - 0.2", etc. I originally hit this point based on an unexpected sign flip in when updating parameter values.  In the code, this meant adding a negative sign in rotate_spectrum (and calc_rot_offset). Do you have any reason to disprefer doing this? I can switch it all back if theres a reason this update doesn't make sense somehow. 
- rotate_spectrum is now quite different from the original function in NDSP. In terms of scoping, I would probably suggest removing that function from NDSP? But if we really want to keep it, we could update it there, similarly to here. 

ToDo: we should add some notes on the derivations that perform these transformations. I'm gonna leave that to @rdgao if possible :). 

@ArcadeShrimp : as a user of this kind of functionality, do you want to eyeball this and check that it makes sense to you? You don't have to worry about the implementation details / math - but just check if this seems useful from a user-facing perspective. 

Note that an open idea in #109 is also to have an option to simulate spectra with specified rotation points directly in the gen_* functions. This is doable (I think, in a clean way) - but not added, and could be. 
^Edit: this idea seems to get a little complicated in terms of dealing with also with generating knee spectra - so will pause on that idea until we figure that part out.

rdgao commented:
- It's very unclear right now how applying a rotation like we do affects the knee parameter, and the interpretation of it thereafter. We're essentially multiplying a `f**(delta)` to the denominator, which tacks onto `k` a frequency dependence that was not previously there if you just generate a Lorentzian spectra with knee. Further, that results in a rotation of the low-freq parts of the Lorentzian as well, leaving you with two disjoint (non-zero) slopes. I'm not sure how the math for out for analytically computing `k`, I'm even more unsure about what `k` means in that case. Suffice to say, this should probably be discussed at a future point.
- neurodsp needs the rotation because that's how it currently generates the colored noise, so we can either just paste over the same function from here, or wrap that into the simulation function itself since it's not used anywhere else (I don't think?)
- yep, happy to write out the math somewhere. Where should we put that, like in one of the tutorial notebooks?

rdgao commented on fooof/synth/transform.py:
```diff
@@ -14,33 +16,124 @@ def rotate_spectrum(freqs, power_spectrum, delta_f, f_rotation):
-        Change in power law exponent to be applied.
-        Positive is counterclockwise rotation (flatten).
-        Negative is clockwise rotation (steepen).
+    delta : float
```
maybe we should name it delta_chi or delta_exp or something? delta by itself is kinda generic

rdgao commented on fooof/synth/transform.py:
```diff
@@ -14,33 +16,124 @@ def rotate_spectrum(freqs, power_spectrum, delta_f, f_rotation):
+    ----------
+    power_spectrum : 1d array
+        Power values of the spectrum that is to be translated.
+    delta : float
```
yeah, especially since there are multiple functions, where delta is used and means different things, it'd be easier to save the user from having to interpret what delta means from the context of the function.

rdgao commented:
other than the two comments above re: delta, looks good to me!

TomDonoghue commented:
Good point on the `delta` variable - that wasn't super clear. I've updated them. 

So yeh, it seems the interaction of rotations with knees is a bit of a thing, so for know I've just added a note / warning that this rotation approach should only be applied to 'fixed' spectra. (Is it possible to subtract the knee out, rotate a spectrum, and add it back in? Seems like that might be a possible solution here. Not required for this PR though). 

For the math: maybe just write it out in a note in the 'Notes' section? This keeps it more attached to the code (and could be reprinted in a tutorial, but you wouldn't need to go there if you just want a quick reference). This means the math would get printed in the API section of the docsite, and you can actually write in latex in the docs, to print as pretty equations on the site, if you want.
